File: 프로그래머스/1/42840. 모의고사/모의고사.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>
#include <algorithm>

using namespace std;

vector<int> solution(vector<int> answers) 
{
    vector<int> vec1 = {1, 2, 3, 4, 5};
    vector<int> vec2 = {2, 1, 2, 3, 2, 4, 2, 5};
    vector<int> vec3 = {3, 3, 1, 1, 2, 2, 4, 4, 5, 5};
    
    int count1 = 0, count2 = 0, count3 = 0;

    for (size_t idx = 0; idx < answers.size(); ++idx) 
    {
        if (answers[idx] == vec1[idx % vec1.size()]) count1++;
        if (answers[idx] == vec2[idx % vec2.size()]) count2++;
        if (answers[idx] == vec3[idx % vec3.size()]) count3++;
    }
    
    // 점수를 쌍으로 저장
    vector<pair<int, int>> scores = 
    {
        {count1, 1},
        {count2, 2},
        {count3, 3}
    };

    // 점수(내림차순) 기준으로 정렬, 동점일 경우 키(오름차순) 기준으로 정렬
    sort(scores.begin(), scores.end(), [](const pair<int, int>& a, const pair<int, int>& b) {
        if (a.first == b.first) return a.second < b.second;
        return a.first > b.first;
    });

    // 최고 점수를 가진 학생 추출
    vector<int> result;
    int highestScore = scores[0].first;

    for (const auto& score : scores) 
    {
        if (score.first == highestScore) 
        {
            result.push_back(score.second);
        }
    }

    return result;
}

File: 프로그래머스/1/42840. 모의고사/모의고사.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution 
{
    public int[] solution(int[] answers) 
    {
        int[] f = new int[] {1, 2, 3, 4, 5};
        int[] s = new int[] {2, 1, 2, 3, 2, 4, 2, 5};
        int[] t = new int[] {3, 3, 1, 1, 2, 2, 4, 4, 5, 5};
        
        int[] scores = new int[3];
        for(int i=0; i<answers.Length; i++)
        {
            if(f[i%5] == answers[i])
            {
                scores[0]++;
            }
            if(s[i%8] == answers[i])
            {
                scores[1]++;
            }
            if(t[i%10] == answers[i])
            {
                scores[2]++;
            }
        }
        
        int max = scores.Max();
        
        List<int> result = new List<int>();
        for(int i = 0; i < 3; i++)
        {
            if(scores[i] == max)
            {
                result.Add(i + 1);
            }
        }
        
        return result.ToArray();
        
    }
}